pymatgen/io/vasp/sets.py

```diff
@@ -458,8 +458,21 @@ def incar(self) -> Incar:
         if self.use_structure_charge:
             incar["NELECT"] = self.nelect
 
-        if np.product(self.kpoints.kpts) < 4 and incar.get("ISMEAR", 0) == -5:
-            incar["ISMEAR"] = 0
+        # Ensure adequate number of KPOINTS are present for the tetrahedron
+        # method (ISMEAR=-5). If KSPACING is in the INCAR file the number
+        # of kpoints is not known before calling VASP, but a warning is raised
+        # when the KSPACING value is > 0.5 (2 reciprocal Angstrom).
+        # An error handler in Custodian is available to
+        # correct overly large KSPACING values (small number of kpoints)
+        # if necessary. 
+        if "KSPACING" not in self.user_incar_settings.keys():
+            if np.product(self.kpoints.kpts) < 4 and incar.get("ISMEAR", 0) == -5:
```

Comment:
Curious, why is this `np.product(self.kpoints.kpts)` and not `len(...)`?

Reply:
If the user is relying on automatic k-point generation then `kpoints.kpts` will just be the mesh, e.g.., `[[4, 4, 4]]`.

This should probably be updated to handle the case where the list of k-points is supplied explicitly?

Reply:
I'm not sure @mkhorton ; that's what was there previously

Reply:
Ah, hence the product previously ... yes, agreed. To handle this robustly I think we'd have to both explicitly test number of k-points from the mesh if the user is using a mesh (testing `np.product(...) < 4` doesn't seem like it'd cover all cases? doesn't spglib have a function to give k-points from a mesh?) and then also for the length of explicit k-point lists as you mention.

In either case, just changing ISMEAR=-5 to ISMEAR=0 may not be the right call here, it may be better to warn the user and/or up the k-point density, it seems like the right action would be context dependent.

Reply:
I'd be happy to refine this scheme but would prefer to do so under a separate PR, since it's not directly related to KSPACING. If you guys tell me what you think the logic should be I'll be glad to implement it.